extensions/mongodb-client/runtime/src/main/java/io/quarkus/mongodb/runtime/graal/BsonSubstitutions.java

```diff
@@ -0,0 +1,106 @@
+package io.quarkus.mongodb.runtime.graal;
+
+import java.text.ParsePosition;
+import java.text.SimpleDateFormat;
+import java.util.Date;
+import java.util.Locale;
+
+import org.bson.json.JsonParseException;
+import org.bson.json.JsonReader;
+
+import com.oracle.svm.core.annotate.Substitute;
+import com.oracle.svm.core.annotate.TargetClass;
+
+import io.quarkus.arc.impl.Reflections;
+
+public final class BsonSubstitutions {
+}
+
+@TargetClass(JsonReader.class)
+final class JsonReaderSubstitution {
+    @Substitute
+    private long visitISODateTimeConstructor() {
+        Util.jsonReaderVerifyToken("LEFT_PAREN", this);
+        // verifyToken(JsonTokenType.LEFT_PAREN);
+
+        Object token = Util.jsonReaderPopToken(this);
+        // JsonToken token = popToken();
+        Object tokenType = Util.jsonTokenGetType(token);
+        if (tokenType == Util.jsonTokenType("RIGHT_PAREN")) {
+            return new Date().getTime();
+        } else if (tokenType != Util.jsonTokenType("STRING")) {
+            throw new JsonParseException("JSON reader expected a string but found '%s'.", Util.jsonTokenGetValue(token));
+            // throw new JsonParseException("JSON reader expected a string but found '%s'.", token.getValue());
+        }
+
+        Util.jsonReaderVerifyToken("RIGHT_PAREN", this);
+        // verifyToken(JsonTokenType.RIGHT_PAREN);
+        String[] patterns = { "yyyy-MM-dd", "yyyy-MM-dd'T'HH:mm:ssz", "yyyy-MM-dd'T'HH:mm:ss.SSSz" };
+
+        SimpleDateFormat format = new SimpleDateFormat(patterns[0], Locale.ENGLISH);
+        ParsePosition pos = new ParsePosition(0);
+        String s = Util.jsonTokenGetValueString(token);
+        // String s = token.getValue(String.class);
+
+        if (s.endsWith("Z")) {
+            s = s.substring(0, s.length() - 1) + "GMT-00:00";
+        }
+
+        for (final String pattern : patterns) {
+            format.applyPattern(pattern);
+            format.setLenient(true);
+            pos.setIndex(0);
+
+            Date date = format.parse(s, pos);
+
+            // Substitution: skip checking length, return as soon as a date is available
+            if (date != null) {
```

Comment:
The original condition in the Java driver looks like this:

    date != null && pos.getIndex() == s.length()

The second condition is required for correctness.  Otherwise, a date string like `2013-10-04T12:07:30.443Z`
will match the first pattern (`yyyy-MM-dd`) with characters still left over, and will return a date this is stripped of the time component contained in the provided date string.  This is not correct.

Reply:
Ok. Let's see if your suggestion in the issue works for the user instead.